spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java

```diff
@@ -0,0 +1,136 @@
+    @Override
+    public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+        this.publisher = applicationEventPublisher;
+    }
+
+    @Override
+    public boolean isAutoStartup() {
+        return true;
+    }
+
+    @Override
+    public void stop(Runnable callback) {
+        this.stop();
+        callback.run();
+    }
+
+    @Override
+    public void start() {
+        if (this.running.compareAndSet(false, true)) {
+            this.watchFuture = this.taskScheduler.scheduleWithFixedDelay(this::watchConfigKeyValues,
+                    this.properties.getWatch().getDelay());
+        }
+    }
+
+    @Override
+    public void stop() {
+        if (this.running.compareAndSet(true, false) && this.watchFuture != null) {
+            this.watchFuture.cancel(true);
+        }
+    }
+
+    @Override
+    public boolean isRunning() {
+        return this.running.get();
+    }
+
+    @Override
+    public int getPhase() {
+        return 0;
+    }
+
+    public void watchConfigKeyValues() {
+        if (this.running.get()) {
+            String prefix = StringUtils.hasText(properties.getPrefix()) ? properties.getPrefix() + "*" : "*";
+            List<KeyValueItem> keyValueItems = configOperations.getKeys(prefix, properties.getLabel());
+
+            if (keyValueItems == null || keyValueItems.isEmpty()) {
+                return;
+            }
+
+            LinkedHashMap<String, String> newKeyEtagMap = keyValueItems.stream()
+                    .collect(Collectors.toMap(KeyValueItem::getKey, KeyValueItem::getEtag,
+                            (v1, v2) -> v1, LinkedHashMap::new));
+            if (firstTime) {
+                keyNameEtagMap.putAll(newKeyEtagMap);
+                firstTime = false;
+                return;
+            }
+
+            Optional<String> changedKey = newKeyEtagMap.entrySet().stream()
+                    .filter(e -> mapNotInclude(keyNameEtagMap, e.getKey(), e.getValue()))
+                    .map(e -> e.getKey())
+                    .findFirst();
+
+            if (!firstTime && changedKey.isPresent()) {
+                LOGGER.trace("Some key matching {} is updated, will send refresh event.", prefix);
+                keyNameEtagMap.clear();
+                keyNameEtagMap.putAll(newKeyEtagMap);
+                RefreshEventData eventData = new RefreshEventData(prefix);
+                publisher.publishEvent(new RefreshEvent(this, eventData, eventData.getMessage()));
+            }
+        }
+    }
+
+    private boolean mapNotInclude(LinkedHashMap<String, String> map, String key, String value) {
+        return !map.containsKey(key) || (map.get(key) != null && !map.get(key).equals(value));
+    }
+
+    /**
+     * For each refresh, multiple etags can change, but even one etag is changed, refresh is required.
+     */
+    class RefreshEventData {
+        private static final String MSG_TEMPLATE = "Some key matching %s has been updated since last check.";
```

Comment:
1. Should be `some keys`
2. My suggestion is including which keys are actually changed to help troubleshooting.

Reply:
In the [RefreshEventListener](https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/endpoint/event/RefreshEventListener.java#L51), the changed keys will be printed, as in here, I will publish an event without checking all keys change status, so the changed keys information is not available here.